test/apply-formation.js

```diff
@@ -0,0 +1,79 @@
+const nock = require('nock');
+const expect = require('chai').expect;
+const proxyquire = require('proxyquire');
+const applyFormation = proxyquire('../src/apply-formation', {
+  'promise-poller': {
+    default: async (config) => {
+      while (!await config.taskFn().then(() => true).catch(_err => false)) {}
+      return true;
+    }
+  }
+});
+
+describe('apply-formation', function () {
+  let success;
+  before(function () {
+    success = nock('https://api.heroku.com')
+      .patch('/apps/exampleApp/formation')
+      .reply(200)
+      .get('/apps/exampleApp/dynos')
+      .times(3)
+      .reply(200, [{ type: 'web', state: 'starting' }])
+      .get('/apps/exampleApp/dynos')
+      .reply(200, [{ type: 'web', state: 'up' }]);
+  });
+  after(function () {
+    nock.cleanAll();
+  });
+  it('requests the formation and waits for it to apply', async function () {
+    await applyFormation(this.heroku, 'exampleApp', [{ type: 'web', quantity: 1 }]);
+    expect(success.isDone()).to.equal(true);
+  });
+
+  describe('multiple dyno types', function () {
```

Comment:
why create another `describe` so that you can have a different `before` for a single `it`?

I'd rather move the `const success = ...` into each test and just have 2 simple `it('should do a thing...')` tests

More `describe` blocks only really makes sense if you have a single `before` block of code you want to apply to many test cases...

```
apply-formation
  ✓ requests the formation and waits for it to apply
  multiple dyno types
    ✓ requests the formation and waits for it to apply

check-formation
  returns true
    ✓ if app has requested formation
  returns false
    ✓ if app does not have requested formation
  returns false - multiple dyno types
    ✓ if app does not have requested formation
```
becomes
```
apply-formation
  ✓ should apply a single dyno formation
  ✓ should apply a multi-dyno formation

check-formation
  ✓ should return true
  ✓ should return false if the app does not have the requested single dyno formation
  ✓ should return false if the app does not have the requested multi-dyno formation
```

Reply:
yup, fair points.  All updated.